src/simplicial-hott/06-2cat-of-segal-types.rzk.md

```diff
@@ -440,3 +501,205 @@ the "Gray interchanger" built from two commutative triangles.
+  : is-contr (fiber-postcomp-Π-ext I ψ ϕ A B a f β)
+  :=
+  is-contr-equiv-is-contr'
+  ( fiber-postcomp-Π-ext I ψ ϕ A B a f β)
+  ( fiber-family-ext I ψ ϕ A B a f β)
+  ( equiv-fiber-postcomp-Π-ext-fiber-family-ext I ψ ϕ A B a f β)
+  ( is-contr-fiber-family-ext-contr-fib I ψ ϕ A B a f β family-equiv-f)
+
+#def is-equiv-postcomp-Π-ext-is-equiv uses (weakextext extext)
+  ( I : CUBE)
+  ( ψ : I → TOPE)
+  ( ϕ : ψ → TOPE)
+  ( A B : ψ → U)
+  ( a : (t : ϕ) → A t)
+  ( f : (t : ψ) → (A t → B t))
+  ( family-equiv-f : (t : ψ) → is-equiv (A t) (B t) (f t))
+  : is-equiv
+    ( (t : ψ) → A t [ϕ t ↦ a t])
+    ( (t : ψ) → B t [ϕ t ↦ f t (a t)])
+    ( postcomp-Π-ext I ψ ϕ A B a f)
+  :=
+  is-equiv-is-contr-map
+  ( (t : ψ) → A t [ϕ t ↦ a t])
+  ( (t : ψ) → B t [ϕ t ↦ f t (a t)])
+  ( postcomp-Π-ext I ψ ϕ A B a f)
+  ( \ β
+    → is-contr-fiber-postcomp-Π-ext-is-equiv-fam I ψ ϕ A B a f β family-equiv-f)
+```
+
+Using this result we can show that `#!rzk ap-hom` is an equivalence when f is an
+equivalence.
+
+```rzk
+#def is-equiv-ap-hom-is-equiv uses (weakextext extext)
+  ( A B : U)
+  ( f : A → B)
+  ( is-equiv-f : is-equiv A B f)
+  ( x y : A)
+  : is-equiv (hom A x y) (hom B (f x) (f y)) (ap-hom A B f x y)
+  :=
+  is-equiv-postcomp-Π-ext-is-equiv 2 Δ¹ ∂Δ¹
+  ( \ _ → A) ( \ _ → B)
+  ( \ t → recOR (t ≡ 0₂ ↦ x , t ≡ 1₂ ↦ y))
+  ( \ _ → f) ( \ _ → is-equiv-f)
+```
+
+More precicely:
+
+```rzk
+#def fiber-ap-hom
+  ( A B : U)
+  ( x y : A)
+  ( f : A → B)
+  ( β : hom B (f x) (f y))
+  : U
+  :=
+  fib (hom A x y) (hom B (f x) (f y)) (ap-hom A B f x y) (β)
+
+#def is-contr-fiber-ap-hom-is-equiv uses (weakextext extext)
+  ( A B : U)
+  ( f : A → B)
+  ( is-equiv-f : is-equiv A B f)
+  ( x y : A)
+  ( β : hom B (f x) (f y))
+  : is-contr (fiber-ap-hom A B x y f β)
+  :=
+  is-contr-fiber-postcomp-Π-ext-is-equiv-fam 2 Δ¹ ∂Δ¹
+  ( \ _ → A) ( \ _ → B)
+  ( \ t → recOR (t ≡ 0₂ ↦ x , t ≡ 1₂ ↦ y))
+  ( \ _ → f)
+  ( β)
+  ( \ _ → is-equiv-f)
+```
+
+We can also define a retraction of `#!rzk ap-hom` directly.
+
+```rzk
+#def retraction-ap-hom-retraction uses (funext)
+  ( A B : U)
+  ( f : A → B)
+  ( (r, ρ) : has-retraction A B f)
+  ( x y : A)
+  : (hom B (f x) (f y)) → (hom A x y)
+  :=
+  comp
+  ( hom B (f x) (f y))
+  ( hom A (r (f x)) (r (f y)))
+  ( hom A x y)
+  ( transport (A → A) (\ g' → (hom A (g' x) (g' y)))
+    ( comp A B A r f)
+    ( identity A)
+    ( eq-htpy funext A (\ x' → A) (comp A B A r f) (identity A) ρ))
+  ( ap-hom B A r (f x) (f y))
+
+#def has-retraction-ap-hom-retraction uses (funext)
+  ( A B : U)
+  ( f : A → B)
+  ( (r, ρ) : has-retraction A B f)
+  ( x y : A)
+  : has-retraction (hom A x y) (hom B (f x) (f y)) (ap-hom A B f x y)
+  :=
+  ( retraction-ap-hom-retraction A B f (r, ρ) x y
+    , \ α →
+      apd (A → A) (\ g' → (hom A (g' x) (g' y)))
+      ( comp A B A r f)
+      ( identity A)
+      ( \ g' → ap-hom A A g' x y α)
+      ( eq-htpy funext A (\ x' → A) (comp A B A r f) (identity A) ρ))
+```
```

Comment:
I think it should be possible to avoid `funext` here.

For example, instead of doing transport over `A -> A` you can do transport pointwise over `prod A A` along the two paths `r (f x) = x` and `r (f y) = y`

Reply:
In fact, if you are willing to use `extext` (which you need anyway), then this is just a special case of `has-retraction-extensions-has-retraction`

Reply:
Nice, thanks! I have changed this proof to use `has-retraction-extensions-has-retraction` instead.